pkg/query/types.go

```diff
@@ -0,0 +1,95 @@
+package query
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"iter"
+	"strings"
+
+	"github.com/authzed/spicedb/pkg/datastore"
+	"github.com/authzed/spicedb/pkg/tuple"
+)
+
+type (
+	// Relation is the basic unit of connection
+	Relation = tuple.Relationship
+	// RelationSeq is the intermediate iter closure that any of the planning calls return.
+	RelationSeq iter.Seq2[Relation, error]
+)
+
+var ErrUnimplemented = errors.New("feature not yet implemented")
+
+// Plan is the external-facing notion of a query plan. These follow the general API for
+// querying anything in the database as well as describing the plan.
+type Plan interface {
+	// Check tests if, for the underlying set of relationships (which may be a full expression or a basic lookup, depending on the iterator)
+	// any of the `resourceIDs` are connected to `subjectID`.
+	// Returns the sequence of matching relations, if they exist, at most `len(resourceIDs)`.
+	Check(ctx *Context, resourceIDs []string, subjectID string) (RelationSeq, error)
```

Comment:
I wonder if this is just a `FilteredIterSubjects`?

Reply:
Kind of, but then so is a full table scan 🙃 

The only reason to have a `[]string` for resources is because datastore lets us use it -- it's completely reasonable to make Check do a single check, as an alternative (returning a Relation because we don't know if there's caveats)

Reply:
I meant more the name; passing in the slice makes sense